### PR TITLE
Finished BP Onboarding 3! Frontend styling fundamentals. 

### PR DESCRIPTION
**Pull request by andi-halim**
## In this PR

Briefly describe what's in this PR.

### 🖋  What I learned
- deeper understanding of typescript
        - enums
        - props
- flexbox
- button formatting

### 🤳 Screenshots (if applicable)
image
flexboxfroggy-1011

### Pop Quiz 

1. What's the difference between padding and margins from the CSS Box Model? Define both.

Padding clears an area around the content while margin clears an area outside the border (a screenshot is shown below that helped me understand it). They are similar in that they are both transparent. 
image

2. What's the difference between align-items and justify-content in CSS FlexBox? Define both. 

Both justify-content and align-items in CSS FlexBox are used to align flex items. The difference is that justify-content aligns flex items along the horizontal axis while align-items aligns flex items along the vertical. 

### 🐞 Questions

Is the best way to get CSS properties from Figma to right-click then "Copy as CSS"/"Copy Properties"? I was a little confused as to how to get styling from the Figma since when I just right-clicked the objects, there wasn't much information. Should I be copy-pasting properties from every object? Feel like there's a better way...

image

I'm still really new to TypeScript and just want to ask for some clarification as to what it is and what are its advantages/differences in comparison to Javascript. Thanks guys!!

### CODE REVIEW

File: src/Button.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+// HINT: Use this in ButtonProps to determine how the button text is styled
+export enum ButtonTextType {
+  white="button-text-white",
+  green="button-text-green",
+  grey="button-text-grey",
+}
+
+interface ButtonProps {
+    text: string;
+    buttonType: ButtonType;
```
Comment (micahtyong) [inline src/Button.tsx:21]: Good work working with enums. Here's an easier way: rather than making those enumerations map keys to strings that correspond to the css selector, use `keyof typeof ButtonType` instead where ButtonType might look something like 
`export enum ButtonType {
  "button-green",
  "button-white",
  "button-clear",
  "button-clearAlt"
}`

**Comment by micahtyong** [comment]
Hey Andi! Clean work with this PR. I'll let Nick answer your first question, but to answer the second, 

> I'm still really new to TypeScript and just want to ask for some clarification as to what it is and what are its advantages/differences in comparison to Javascript.

The main motivation behind TypeScript was that it's statically typed. From [this article by Paul Chiusano](http://pchiusano.github.io/2016-09-15/static-vs-dynamic.html), here are some of the advantages of statically typed languages: 

- A large class of errors are caught, earlier in the development process, closer to the location where they are introduced. For example, in the button activity, TypeScript would complain if it received props input other than "green", "white", etc. So, as a result, we catch errors earlier on as we code rather than waiting to discover it at runtime! If you've taken 61B, they analyze the difference between compile-time and runtime quite well. 
- The types guide development and can even write code for you—the programmer ends up with less to specify. 
- One can refactor with greater confidence, since a large class of errors introduced during refactoring will end up as type errors.
- Static types can ease the mental burden of writing programs, by automatically tracking information the programmer would otherwise have to track mentally in some fashion.
- For Blueprint in particular, we normally hand-off our application to a developer after we finish, and TypeScript makes it way easier for a new developer to get acquainted with a large codebase,  
- There's a whole lot more. 

**TLDR**: For larger-scale programs and programs that are worked on for more people (e.g. ones worked on at Blueprint), TypeScript has loads of benefits that make development more seamless and scalable in the long run. 

_This is a personal opinion, but I think you are going to see more and more of industry use TypeScript in the next few years because of the advantages stated earlier._
